### ISSUE TITLE
Permanent deletion of product returns true even tough product is not permanently deleted

### ISSUE THREAD

**Issue by Dekadinious** [woothemes/woocommerce#33674]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

Case: Send a product to trash using `$product->delete();` and follow that up with `$product->delete(true);` at a later stage without first fetching the product again using `wc_get_product();`. `$product->delete(true);` will return true, but the product will still be present in the cart. If you do `wc_get_product` in between the two methods, the product will be permanently deleted.

### Expected behavior

The method should return false, or the product should be permanently deleted.

### Actual behavior

The method returns true and the product is not permanently deleted.

### Steps to reproduce

```
$product = wc_get_product(738800);
\WP_CLI::log(var_export($product->delete(), 1));
\WP_CLI::log(var_export($product->delete(true), 1));
```
Produces `true` `true`, product still present in trash.

```
$product = wc_get_product(738800);
\WP_CLI::log(var_export($product->delete(), 1));
$product = wc_get_product(738800);
\WP_CLI::log(var_export($product->delete(true), 1));
```
Produces `true` `true`, product gone from trash.

### WordPress Environment

Sorry for not providing it at this time.

### Isolating the problem

- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by barryhughes** [comment]
Thanks for reporting this, I can see the same thing.

The underlying problem is that once you call `$product->delete()` the `id` property is set to `(int) 0`. Further operations respect this (which is why the second call to `->delete()` doesn't do anything: it's trying to delete product `0` at that point).

This probably isn't very clear, and could definitely be confusing. I'm a little unsure as to priority, however. Is there any other context you can provide about the work you were doing when you encountered this to help us decide?

**Comment by Dekadinious** [comment]
@barryhughes 

Of course! I was working on a product sync plugin and found that I needed to free the SKU for use when converting a simple product to a variation. So what I am doing is first freeing the SKU by deleting the simple product (to trash), then I am creating a new variation with all the data from the simple product and attaching it to the correct mother. If everything goes well I am permanently deleting the simple product (this happens in the same run, and therefore the problem happens). If something does not go according to plan, the simple product still exists in the trash with it's original SKU so that it can be manually untrashed by a human without looking up and setting the SKU. That's why I don't just set the SKU to something else or blank instead of deleting the product.

**Comment by barryhughes** [comment]
Great, thank you!

In that case I'll set this to low priority, as it seems like this will not impact the majority of users—and it should also be possible to workaround it by acquiring a new instance of the product when the time comes to fully delete it (or to restore it):

```php
// Trash.
wc_get_product( $product_id )->delete();

// Completely delete.
wc_get_product( $product_id )->delete( true );
```

**Comment by Dekadinious** [comment]
That's exactly how I solved it, yes :)

**Comment by pmcpinto** [comment]
Thanks for the report.

I know this has been open for a very long time, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

I'd like to get an engineer's eyes on this before we make a call, so I've flagged it for developer feedback. Someone will take a proper look at the repro steps and how it's behaving now.
